### DISCUSSION TITLE
Having issue integrating bloom filter support using JedisPooled

### DISCUSSION THREAD

**Discussion by Ranjeet116966** [xetorthio/jedis#4028] [Q&A]
I am currently working on integrating Bloom filter support into one of our features. For this, I’m using JedisPooled for the Bloom filter-related operations, as I believe Jedis itself doesn’t support these methods. After deploying these changes to production, I'm encountering an error `("Unexpected end of stream" / JedisConnectionException)` on every second request, causing those requests to fail. I’m unsure whether this issue is due to the size of the connection pool or because the timeouts are set too low for Jedis.

I am using JedisPooled in the same way it’s mentioned in the README file, where I create an instance of JedisPooled like this and pass the custom config I have:

`JedisPooled jedis = new JedisPooled("localhost", 6379);`

Then, I issue commands on this` jedis` instance.

```
Jedis version: 4.2.2
Redis version: 7.2.4
```

Here is my current Jedis configuration:

```
connection_timeout: 200
read_timeout: 3000
max_total: 40
max_idle: 5
max_wait: 100
```
Can someone please take a look and help me understand if i am missing something, and if somehow i can get rid of these errors TIA...

**Comment by sazzad16** [comment]
@Ranjeet116966 Provide a short code that reproduces the issue and add the full stacktrace.

**Comment by Ranjeet116966** [comment]
Here is the full stack trace, will provide the code snippet in sometime, Actually i am integrating bloom filter in some gem, that we consume on some other repo, all the config that we need to initialise the connection pool would be passed from consumer end, so will have to create a short code file to replicate the same behaviour

` ERROR jruby.app- Connection handling error: Unexpected end of stream.
redis.clients.jedis.util.RedisInputStream.ensureFill(redis/clients/jedis/util/RedisInputStream.java:205)
redis.clients.jedis.util.RedisInputStream.readByte(redis/clients/jedis/util/RedisInputStream.java:46)
redis.clients.jedis.Protocol.process(redis/clients/jedis/Protocol.java:126)
redis.clients.jedis.Protocol.read(redis/clients/jedis/Protocol.java:192)
redis.clients.jedis.Connection.readProtocolWithCheckingBroken(redis/clients/jedis/Connection.java:316)
redis.clients.jedis.Connection.getOne(redis/clients/jedis/Connection.java:298)
redis.clients.jedis.Connection.executeCommand(redis/clients/jedis/Connection.java:123)
redis.clients.jedis.executors.DefaultCommandExecutor.executeCommand(redis/clients/jedis/executors/DefaultCommandExecutor.java:24)
redis.clients.jedis.UnifiedJedis.executeCommand(redis/clients/jedis/UnifiedJedis.java:148)
redis.clients.jedis.UnifiedJedis.bfExists(redis/clients/jedis/UnifiedJedis.java:3923)
java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:498)
org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:470)
org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:328)`

**Comment by Ranjeet116966** [comment]
Here is the basic code snippet you can refer to, this is how i have also implemented, and i am getting the above error after deploying this code to prod 

```
require 'java'

require 'redis.clients:jedis-4.2.2.jar'
require 'org.slf4j:slf4j-api-2.0.13.jar'
require 'org.apache.logging.log4j:log4j-slf4j2-impl-2.23.1.jar'

java_import 'redis.clients.jedis.JedisPooled'
java_import 'redis.clients.jedis.JedisPoolConfig'
java_import 'redis.clients.jedis.bloom.RedisBloomCommands'

connection_timeout = 200
read_timeout = 3000

def initialize_pool(connection_timeout, read_timeout)
  jedis_pool_config = JedisPoolConfig.new
  jedis_pool_config.setMaxTotal(5)
  jedis_pool_config.setMaxIdle(5)
  jedis_pool_config.setMaxWaitMillis(100)
  jedis_pool_config.setBlockWhenExhausted(true)

  JedisPooled.new(
    jedis_pool_config,
    java.net.URI.create("redis://localhost:6380"),
    connection_timeout,
    read_timeout
  )
end

def execute_bloom_filter_command(pool)
  pool.bfAdd('bloom_filter_key', 'item')
  puts pool.bfExists('bloom_filter_key', 'item')
end

pool = initialize_pool(connection_timeout, read_timeout)
execute_bloom_filter_command(pool)
```

**Comment by sazzad16** [comment]
@Ranjeet116966 Couldn't reproduce your issue. Can't even guess what could go wrong within localhost.

- What's the timeout in Redis server? I.g. redis-cli ==> `CONFIG GET timeout` or Jedis ==> `configGet("timeout")`.
- Could you try increasing the connection timeout? I.e. as high the read_timeout. (Seems doesn't matter in case, but just to be safe.)
- Could you try an updated Jedis version? If the latest version is not possible, at least Jedis 4.4.8?

**Comment by Ranjeet116966** [comment]
Reproducing the error is difficult because it's not very frequent, some of the request are passing and some request are failing so error is very inconsistent in my testing env as well. 

- Redis timeout is set to 0 i think 

```
CONFIG GET timeout
1) "timeout"
2) "0"
```

Will try increasing the` connect_timeout `and if that doesn't help will also upgrade the jedis version, Did we change anything related to this issue, i didn't find anything like that in release notes

**Reply by sazzad16** [reply]
> ... some of the request are passing and some request are failing ...

What's the ratio of passing and failing?

**Reply by Ranjeet116966** [reply]
it's almost 50:50 so basically half of the total requests are failing with above error

**Comment by Ranjeet116966** [comment]
Seems like none of the above solution is working for me, have tried all of that and still seeing the same issue. Let me know if there is anything that can help TIA!!

**Comment by Ranjeet116966** [comment]
could this be related to above error - https://github.com/redis/jedis/issues/3864 ?

**Reply by sazzad16** [reply]
No. It doesn't seem so.

**Comment by sazzad16** [comment]
1.
> I'm encountering an error `("Unexpected end of stream" / JedisConnectionException)` on every second request, ...
2.
> Reproducing the error is difficult because it's not very frequent, ...

Humbly speaking, I find the above two statements conflicting.

**Reply by Ranjeet116966** [reply]
by not very frequent i mean it's not consistent because as i said 50% request are still passing

**Comment by Ranjeet116966** [comment]
Hi, a quick update on this: It seems that changing the `client-output-buffer-limit` for `pubsub ` to 0 has resolved the issue. However, I'm not entirely sure why this fixed it, as I’m not using the pubsub client for my specific use case.

```
CONFIG get client-output-buffer-limit:

"client-output-buffer-limit"
"normal 0 0 0 slave 268435456 67108864 60 ¸ 0 0 0"
```

**Comment by Ranjeet116966** [comment]
Hi @sazzad16,

Is there a significant difference between `JedisPool` and `JedisPooled` class? I've been using `JedisPooled `for running Bloom filter module commands, but I'm encountering above error. Other Redis commands still use `JedisPool`, and I tried switching all Redis commands to `JedisPooled`, but I'm getting an "Unexpected end of stream" error. The same commands work fine with `JedisPool`, but not with `JedisPooled`.

**Reply by sazzad16** [reply]
No differences as how they function. The difference is in only how you use it.

**Reply by Ranjeet116966** [reply]
Okay so in case of JedisPooled we don't have to manage the connections and object pooling and in case of JedisPool we have to it by ourself, I have provided you a code block earlier that's how currently we are using JedisPooled, and for JedisPool class we have one extra method in the code that get's the resource from the pool and release it once the command is executed with a try catch block. Also i am passing the same config to both the JedisPool and JedisPooled but still running into issues when trying to use JedisPooled.

**Reply by Ranjeet116966** [reply]
Hi @sazzad16 can you please provide some suggestion here, Also we are using` jruby`, Can that be the reason of this different behaviour that we are seeing.

**Reply by sazzad16** [reply]
@Ranjeet116966 Sorry, I could not help this time. Ideally, JedisPool and JedisPooled should have same connection/resource management and processing. From all the information you provided, I could not pinpoint any issue in JedisPooled. I have no other idea about what could be the reason.

As about jruby, may be. But I am not sure.

**Reply by Ranjeet116966** [reply]
Okay thanks @sazzad16 then i will try to use `JedisCluster` class to see if that resolve this issue.

**Reply by ayansaha23** [reply]
did it resolve, I get  error bfReserve is not a member of redis.clients.jedis.JedisPool?

**Reply by sazzad16** [reply]
@ayansaha23 seems to be different issue.

bfReserve is not a method in JedisPool class. You would have to used JedisPooled class for that.

If you issue is not even that, please create a new discussion with more information.

**Reply by ayansaha23** [reply]
yes thanks I got it to work using JedisPooled class

**Reply by Ranjeet116966** [reply]
@ayansaha23 can you please tell what version of jedis library you are using and redis version as well?

**Comment by Ranjeet116966** [comment]
@sazzad16 I tried using JedisCluster, and it worked without any errors. However, my Redis cluster has 3 masters and 3 slaves, but since I'm using only one key for the Bloom filter, I can't effectively utilize the space across those nodes. Is it possible to use the JedisCluster class with a non-sharded Redis setup, and what would be the implications of doing so?

**Reply by ggivo** [reply]
@Ranjeet116966 
No, JedisCluster is designed for Redis Cluster. It will try to load slot distribution on creation and will produce an error when connected to non-cluster env.

Have you tested with more recent Jedis version, as suggested before?
Error from initial issue report `"Unexpected end of stream. "` suggests that connection is forcibly closed while there is ongoing operation.  I also tried to reproduce the mentioned error but without success.  


Jedis : 4.4.8
Redis : docker run -p 6379:6379 redis/redis-stack:7.4.0-v1

```java
    public static void testBloomFilters(String uri) {
        ConnectionPoolConfig poolConfig  = new ConnectionPoolConfig();
        poolConfig.setMaxTotal(5);
        poolConfig.setMaxIdle(5);
        poolConfig.setMaxWaitMillis(100);
        poolConfig.setBlockWhenExhausted(true);

        try (JedisPooled jedis = new JedisPooled(poolConfig, URI.create(uri),100,100);) {

            for (int i = 0; i < 100; i++) {
                jedis.bfAdd("bloom_filter_key", "item");
                boolean resp = jedis.bfExists("bloom_filter_key", "item");
                System.out.println("Response from Redis: " + resp);
            }

        } catch (Exception e) {
            System.err.println("Error connecting to Redis: " + e.getMessage());
        }
    }
```

**Reply by Ranjeet116966** [reply]
@ggivo yes i have tried these versions - `5.1.0`,  `5.2.0` but unfortunately nothing helped !!

**Reply by Ranjeet116966** [reply]
@ggivo one more thing which class should i use to setup the connection pool, right now i see you are using ` ConnectionPoolConfig` and i have used `JedisPoolConfig` and when looking at the codebase seems like both classes are extending `GenericObjectPoolConfig` so which one is recommended approach for configuring connection pool?

**Comment by ggivo** [comment]
`JedisPooled` expects `GenericObjectPoolConfig<Connection>`.  `ConnectionPoolConfig` is an OOB implementation providing defaults settings for the`GenericObjectPoolConfig<Connection>` pool and is used with `JedisPooled`.
  
  Legacy 'JedisPool' class expects `GenericObjectPoolConfig<Jedis> `.  `JedisPoolConfig` is  an OOB implementation providing defaults for the `GenericObjectPoolConfig<Jedis>` pool and is used with  'JedisPool'

**Reply by Ranjeet116966** [reply]
@ggivo so should i use `ConnectionPoolConfig` if i am working with JedisPooled class, could this be a reason for the issue that i am facing, cause right now i am using `JedisPoolConfig`

**Reply by ggivo** [reply]
@Ranjeet116966 
How exactly do you create it? Which version/constructor do you use?
Have not dug into older versions, but in the later one, when constructing `JedisPooled` class it expects  `GenericObjectPoolConfig<Connection>`,  and you should get a compile-time error trying to provide `GenericObjectPoolConfig<Jedis>`

**Reply by Ranjeet116966** [reply]
okay so right now i am using jedis version - 4.2.2

this is how i am reading the config from a yaml file 

````
              basic_cache_config = config['basic_cache']
              jedis_pool_config = JedisPoolConfig.new
              jedis_pool_config.setMaxTotal(basic_cache_config['connection_pool']['max_total'])
              jedis_pool_config.setMaxIdle(basic_cache_config['connection_pool']['max_idle'])
              jedis_pool_config.setMaxWaitMillis(basic_cache_config['connection_pool']['max_wait'])
              jedis_pool_config.setBlockWhenExhausted(true)`
```
          
  and below is the constructor that  i am using and with this setup i did not get any error
  
  ```
` JedisPooled.new(
              jedis_pool_config,
              java.net.URI.create("redis://#{redis_connection[:host]}:#{redis_connection[:port]}"),
              basic_cache_config['connection_timeout'],
              basic_cache_config['read_timeout']
            )`
```

**Reply by ggivo** [reply]
@Ranjeet116966 
Ruby is not type-safe if you try to rewrite this in Java you will get an error at compile time.

Based on your code & latest Jedis code base  it looks like you are using this constructor expecting `"GenericObjectPoolConfig<Connection>"`
```
  public JedisPooled(final GenericObjectPoolConfig<Connection> poolConfig, final URI uri,
      final int connectionTimeout, final int soTimeout) {
    this(poolConfig, uri, connectionTimeout, soTimeout, null, null, null);
  }
```

Try providing the correct poolConfig type.

**Reply by Ranjeet116966** [reply]
@ggivo tried this as well but with no luck, still running into same issue!!

**Reply by Ranjeet116966** [reply]
Is there anything I can adjust in my connection pool configuration? Here’s my current setup:

```
max_total: 10
max_idle: 5
max_wait: 100
```

Also, should I modify the Redis timeouts? Right now, both `timeout` and `tcp-keepalive` are set to 0. The only solution that has worked without any errors so far is using the JedisCluster class with sharded Redis. However, my concern is that I'm only working with a single key, so using a Redis cluster in this case doesn’t really make sense.

**Reply by Ranjeet116966** [reply]
@ggivo Not all requests are failing. When I query the Bloom filter about 100 times, roughly 20 requests fail while the other 80 succeed. I've tested the Bloom filter locally but haven't been able to reproduce the error. The issue only happens when the backend receives calls from clients (Android/web) in the staging or production environment. I'm storing around 900 million SHA-1 hashes in a Bloom filter under a single key.

**Reply by ggivo** [reply]
@Ranjeet116966 
Nothing that I can suggest. If you are not able to reproduce it locally then it might be related to the environment you are running it.
Without being able to reproduce it consistently not able to provide any guidance.

**Reply by Ranjeet116966** [reply]
@ggivo Is it possible to add Bloom filter support to the `JedisPool `class? We're currently using `JedisPool` for all our non-sharded Redis instances, and it has been stable without this error. I also tried migrating from `JedisPool` to `JedisPooled` for our non-sharded Redis setup, but I encountered the same error there as well. Currently, I'm using `JedisPooled` only for Bloom filter operations, while other Redis methods still use the `JedisPool `class.

**Reply by ggivo** [reply]
There are no plans to add Bloom filter support to JedisPool. The issue you’re encountering does not seem to be not reproducible, which makes it difficult to justify changes specifically for JedisPool. Given this, additional effort in that direction might not be the best use of time, as it could lead to similar or different issues down the line.

Going forward, we are focusing on improving UnifiedJedis (and by extension, JedisPooled).

**Reply by Ranjeet116966** [reply]
@ggivo Could you please help me understand the possible reasons for this error? It would help me debug the issue further. I’ve noticed one pattern — I’m making around 1000 requests using a thread pool, with a maximum of 50 concurrent requests to query the Bloom filter. When I start the script, I get an `"Unexpected end of stream"` error, but after about a minute, the errors stop. Could it be that the connection pool, after being idle for a long time, isn’t ready to handle a sudden spike of 1000 requests?

Below is my jedis pool config and one more thing when i am printing the number of idle and active connections, so max number of idle connection that i can see is 5 and max active connection even during the spike 4 only 

```
`      connection_timeout: 200
        read_timeout: 2000
        connection_pool:
          max_total: 10
          max_idle: 5
          max_wait: 100`
```

**Reply by Ranjeet116966** [reply]
also is it possible to use UnifiedJedis class for non sharded redis

**Reply by ggivo** [reply]
I am not able to provide suggestions without having the issue isolated & reliably reproducible.  
It might also be related to the ruby Java interoperability.  For sure, the provided example script (considering it is using the latest Jedis version) is invoking a respective constructor using an object from the wrong type. 

UnifiedJedis is the base class used both for clustered and non-clustered env depending on how it is created/initialized. 
JedisPooled provides a pool of connections toward individual nodes.

**Reply by Ranjeet116966** [reply]
@ggivo  I don't think this is a constructor issue anymore, i am passing the correct constructor now and we are already using jedispool and jediscluster class for our sharded and non sharded redis it's been working fine for us. we are using jruby and this is how we are instantiating `JedisCluster ` and it's working fine in prod env - 

```
` basic_cache_config = config['basic_cache']
          jedis_pool_config = ConnectionPoolConfig.new
          jedis_pool_config.setMaxTotal(basic_cache_config['connection_pool']['max_total'])
          jedis_pool_config.setMaxIdle(basic_cache_config['connection_pool']['max_idle'])
          jedis_pool_config.setMaxWaitMillis(basic_cache_config['connection_pool']['max_wait'])
          jedis_pool_config.setBlockWhenExhausted(true)

          jedis_cluster_nodes = java.util.HashSet.new
          redis_cluster_hosts_ports.each do |redis_cluster_host_port|
            host_and_port = HostAndPort.new(redis_cluster_connection[:host], redis_cluster_connection[:port])
            jedis_cluster_nodes.add(host_and_port)
          end

          JedisCluster.new(
            jedis_cluster_nodes,
            # Timeouts are in milliseconds.
            basic_cache_config['read_timeout'],
            jedis_pool_config
          )`

```


and something similar i am doing for instantiating `JedisPooled ` class so it's doesn't seem to be an env issue because other two classes are working fine, and `JedisCluster ` is also working with bloom filter, the only class that is not working for me intermittently  is  `JedisPooled `

 ```
basic_cache_config = config['basic_cache']
          jedis_pool_config = JedisPoolConfig.new
          jedis_pool_config.setMaxTotal(basic_cache_config['connection_pool']['max_total'])
          jedis_pool_config.setMaxIdle(basic_cache_config['connection_pool']['max_idle'])
          jedis_pool_config.setMaxWaitMillis(basic_cache_config['connection_pool']['max_wait'])
          jedis_pool_config.setBlockWhenExhausted(true)

      

            JedisPooled.new(
              jedis_pool_config,
              java.net.URI.create("redis://#{redis_connection[:host]}:#{redis_connection[:port]}"),
              basic_cache_config['connection_timeout'],
              basic_cache_config['read_timeout']
            )
```


I am suspecting this could be the cause - https://github.com/redis/jedis/issues/1029#issuecomment-475551912 but even setting up redis timeout to 0 and keeping no idle connection on pool is not working i am still seeing few failures.

**Reply by ggivo** [reply]
@Ranjeet116966 
Your latest example still uses the wrong type for **jedis_pool_config**.
There is no constructor in `JedisPooled` that accepts `JedisPoolConfig`. You need to provide `ConnectionPoolConfig`similar to the example with `JedisCluster`,

**Reply by Ranjeet116966** [reply]
@ggivo sorry shared the old code i am using `ConnectionPoolConfig` only but even that is not working for me  - 

```
` basic_cache_config = config['basic_cache']
          jedis_pool_config = ConnectionPoolConfig.new
          jedis_pool_config.setMaxTotal(basic_cache_config['connection_pool']['max_total'])
          jedis_pool_config.setMaxIdle(basic_cache_config['connection_pool']['max_idle'])
          jedis_pool_config.setMaxWaitMillis(basic_cache_config['connection_pool']['max_wait'])
          jedis_pool_config.setBlockWhenExhausted(true)
          jedis_pool_config.setTestWhileIdle(true)

            JedisPooled.new(
              jedis_pool_config,
              java.net.URI.create("redis://#{redis_connection[:host]}:#{redis_connection[:port]}"),
              basic_cache_config['connection_timeout'],
              basic_cache_config['read_timeout']
            )`
```

**Reply by Ranjeet116966** [reply]
@ggivo do you think this could be the reason - I am suspecting this could be the cause - https://github.com/redis/jedis/issues/1029#issuecomment-475551912 but even setting up redis timeout to 0 and keeping no idle connection on pool is not working i am still seeing few failures.

**Comment by Ranjeet116966** [comment]
Hi @garydgregory,

I wanted to tag you in this discussion to check if the issue I’m facing could be related to the Apache Commons Pool library. I noticed that a similar issue was raised in the past: [Jedis Issue #3106](https://github.com/redis/jedis/issues/3106).

Could you confirm if any changes were made to Apache Commons Pool that might be relevant here?

For reference, I am using Apache Commons Pool version 2.11.1.

Thanks in advance!

**Reply by garydgregory** [reply]
Hello @Ranjeet116966 

I can't really say anything aside from try the current version of Commons Pool (2.12.1) and you could also try Commons Collections 4.5.0-M3 if you want to try our Bloom Filter implementation. 

If you think there is an issue in Commons Pool, I'll need a reproducible test case either as a GitHub PR or an attachment to a Jira issue, see https://commons.apache.org/proper/commons-pool/issue-management.html

HTH

**Reply by Ranjeet116966** [reply]
Thanks, @garydgregory, for the response! I'll try out the latest version of Commons Pool and also take a look at the Bloom filter implementation in Commons Collections 4.5.0-M3.

I have a question regarding how this would work with Redis. Currently, I’m using the RedisBloom module, which I install on the Redis server and interact with using the Jedis client. Given this setup, how would Commons Collections' Bloom filter implementation integrate with Redis?

Would there be additional steps needed to make it work in my current environment?

**Reply by garydgregory** [reply]
Hello @Ranjeet116966 

The Apache Commons Collections [Bloom Filter package](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/bloomfilter/package-summary.html) is standalone, so you must do whatever Jedis needs to integrate it.

- Introduction: https://commons.apache.org/proper/commons-collections/bloomFilters/intro.html
- Package Javadoc: https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/bloomfilter/package-summary.html
